### PR TITLE
[donotmerge] new boot from volume

### PR DESCRIPTION
**Pull request by kaccardi**
Bug fixes and new support for booting from volume.

### CODE REVIEW

**Comment by kaccardi** [comment]
Do not merge yet - release cluster not quite ready.

**Comment by markdryan** [comment]
@kaccardi This is a big PR.  Could you summarize what has changed since the old boot for volume #625?  Should we close #625 now?

**Comment by kaccardi** [comment]
this PR passes bat tests on the release cluster now.

**Comment by kaccardi** [comment]
List of changes from the other PR (roughly):
* remove the bulk workload start test from bat tests
* increase the default timeout for commands to complete in bat tests
* correctly update state of volumes when deleting instances with attached volumes.
* list attached volumes with ciao-cli instance list -detail.
* prevent duplicate datastore entries of workload storage rows

File: ciao-controller/internal/datastore/sqlite3db.go [hunk 1]
```diff
@@ -838,14 +838,16 @@ func (ds *sqliteDB) getWorkloadDefaults(ID string) ([]payloads.RequestedResource
 func (ds *sqliteDB) getWorkloadStorage(ID string) (*types.StorageResource, error) {
 	query := `SELECT volume_id, bootable, persistent, size,
 			 source_type, source_id
-		  FROM 	workload_resources
+		  FROM 	workload_storage
```
Comment (rbradford) [inline ciao-controller/internal/datastore/sqlite3db.go]: Could you squash this patch into the first one in the series where this function is added?

File: ciao-controller/controller_test.go [hunk 2]
```diff
@@ -1126,6 +1127,50 @@ func TestGetStorage(t *testing.T) {
 	}
 }
 
+func TestStorageConfig(t *testing.T) {
+	var err error
```
Comment (rbradford) [inline ciao-controller/controller_test.go]: I'm sooper confused. This patch seems to do so much more than the description says. Should it be rolled into another patch? Or separated?

File: ciao-controller/tables/workload_template.csv [hunk 3]
```diff
@@ -4,3 +4,4 @@ eba04826-62a5-48bd-876f-9119667b1487,CNCI,test.yaml,efi,qemu,4e16e743-265a-4bf2-
 ca957444-fa46-11e5-94f9-38607786d9ec,Docker Ubuntu latest,docker-ubuntu.yaml,"",docker,fa7d86d8-fa46-11e5-8493-38607786d9ec,"ubuntu:latest",0
 ab68111c-03a6-11e6-87de-001320fb6e31,Docker Iperf,docker-iperf.yaml,"",docker,b5e696b8-03a6-11e6-a424-001320fb6e31,"mcastelino/iperf:latest",0
 79034317-3beb-447e-987d-4e310a8cf410,Boot Fedora23 from created volume based on image,test.yaml,legacy,qemu,"","", 0
+1cd4c4d9-ff84-4bfb-a371-a220b129977a,Boot Fedora23 from existing volume,test.yaml,legacy,qemu,"","", 0
```
Comment (rbradford) [inline ciao-controller/tables/workload_template.csv]: Why is this one in this patch but the "79034317-3beb-447e-987d-4e310a8cf410" in the previous patch?

File: ciao-cli/volume.go [hunk 4]
```diff
@@ -62,7 +63,8 @@ The add flags are:
 
 func (cmd *volumeAddCommand) parseArgs(args []string) []string {
 	cmd.Flag.StringVar(&cmd.name, "name", "", "Volume name")
-	cmd.Flag.StringVar(&cmd.source, "source", "", "Volume ID to clone from")
+	cmd.Flag.StringVar(&cmd.sourceType, "source_type", "image", "The type of the source to clone from")
```
Comment (rbradford) [inline ciao-cli/volume.go]: Why do we need this, would an image and a volume have the same UUID? Can't we try image first then fallback to volume?

File: ciao-storage/ceph.go [hunk 5]
```diff
@@ -56,6 +56,7 @@ func (d CephDriver) CreateBlockDevice(imagePath *string, size int) (BlockDevice,
 	return BlockDevice{ID: ID}, nil
 }
 
+// CopyBlockDevice will copy an existing volume
 func (d CephDriver) CopyBlockDevice(volumeUUID string) (BlockDevice, error) {
```
Comment (rbradford) [inline ciao-storage/ceph.go]: Could you squash this into the patch that adds this function.

File: ciao-controller/internal/datastore/sqlite3db.go [hunk 6]
```diff
@@ -255,15 +255,15 @@ func (d workloadStorage) Populate() error {
 	}
 
 	for _, line := range lines {
-		workload_id := line[0]
-		volume_id := line[1]
+		workloadID := line[0]
+		volumeID := line[1]
```
Comment (rbradford) [inline ciao-controller/internal/datastore/sqlite3db.go]: I'd prefer to see this squashed into the originating patch.

File: ciao-storage/ceph.go [hunk 7]
```diff
@@ -42,7 +42,8 @@ func (d CephDriver) CreateBlockDevice(imagePath *string, size int) (BlockDevice,
 	// Currently the kernel rdb client only supports layering but in the future more feaures
 	// should be added as they are enabled in the kernel.
 	if imagePath != nil {
-		cmd = exec.Command("qemu-img", "convert", *imagePath, ID)
+		rbdStr := fmt.Sprintf("rbd:rbd/%s:id=%s", ID, d.ID)
+		cmd = exec.Command("qemu-img", "convert", "-f", "qcow2", "-O", "rbd", *imagePath, rbdStr)
```
Comment (rbradford) [inline ciao-storage/ceph.go]: I'm sure i'm starting to sound a like a stuck record. But I think this would be best rolled into previous patches.

File: ciao-controller/tables/workload_template.csv [hunk 8]
```diff
@@ -4,4 +4,3 @@ eba04826-62a5-48bd-876f-9119667b1487,CNCI,test.yaml,efi,qemu,4e16e743-265a-4bf2-
 ca957444-fa46-11e5-94f9-38607786d9ec,Docker Ubuntu latest,docker-ubuntu.yaml,"",docker,fa7d86d8-fa46-11e5-8493-38607786d9ec,"ubuntu:latest",0
 ab68111c-03a6-11e6-87de-001320fb6e31,Docker Iperf,docker-iperf.yaml,"",docker,b5e696b8-03a6-11e6-a424-001320fb6e31,"mcastelino/iperf:latest",0
 79034317-3beb-447e-987d-4e310a8cf410,Boot Fedora23 from created volume based on image,test.yaml,legacy,qemu,"","", 0
-1cd4c4d9-ff84-4bfb-a371-a220b129977a,Boot Fedora23 from existing volume,test.yaml,legacy,qemu,"","", 0
```
Comment (rbradford) [inline ciao-controller/tables/workload_template.csv]: Why is this workload dropped?

**Comment by kaccardi** [comment]
@rbradford why don't I break this into a series of smaller PRs.

File: ciao-storage/ceph.go [hunk 9]
```diff
@@ -42,7 +42,7 @@ func (d CephDriver) CreateBlockDevice(imagePath *string, size int) (BlockDevice,
 	// Currently the kernel rdb client only supports layering but in the future more feaures
```
Comment (tpepper) [inline ciao-storage/ceph.go]: Is this comment still relevant after the switch to qemu-img?

Comment (tpepper) [inline ciao-storage/ceph.go]: Nevermind...there's a later rbd invocation.

File: ciao-storage/ceph.go [hunk 10]
```diff
@@ -43,7 +43,7 @@ func (d CephDriver) CreateBlockDevice(imagePath *string, size int) (BlockDevice,
 	// should be added as they are enabled in the kernel.
 	if imagePath != nil {
 		rbdStr := fmt.Sprintf("rbd:rbd/%s:id=%s", ID, d.ID)
-		cmd = exec.Command("qemu-img", "convert", "-f", "qcow2", "-O", "rbd", *imagePath, rbdStr)
+		cmd = exec.Command("qemu-img", "convert", "-O", "rbd", *imagePath, rbdStr)
```
Comment (tpepper) [inline ciao-storage/ceph.go]: This could squash back into the original switch from rbd to qemu-img.

File: _release/bat/release_test.py [hunk 11]
```diff
@@ -405,10 +405,6 @@ def test_start_all_workloads(self):
         """Start one instance of all workloads"""
         self.failUnless(launch_all_workloads())
 
-    def test_start_all_workloads10(self):
```
Comment (tpepper) [inline _release/bat/release_test.py]: How about we call this "test_start_all_cattle10()" and have that only launch the cattle type workloads and not ones that are going to amount to download image from network'd image store, copy/convert it to a qemu-img, upload to network'd volume store, and boot from volume.  Meaningful cattle aren't going to do a synchronous conversion in their boot flow.

File: ciao-controller/tables/workload_storage.csv [hunk 12]
```diff
@@ -0,0 +1 @@
+79034317-3beb-447e-987d-4e310a8cf410,"",1,1,0,"image",73a86d7e-93c0-480e-9c41-ab42f69b7799
```
Comment (tpepper) [inline ciao-controller/tables/workload_storage.csv]: Would be helpful here if the commit message described what this line means.  I think it means declare that volume uuid 79034317-3beb-447e-987d-4e310a8cf410 will be a copy of the image uuid 73a86d7e-93c0-480e-9c41-ab42f69b7799?  If the first '1' is bootable and the second '1' is persistent, this would be a workload-specific read-write volume and something would enforce that only one instance of this workload would be allowed in order to prevent two instances writing to the same volume?

**Comment by kaccardi** [comment]
all of these commits where submitted elsewhere for review.
